In [345]:
import pandas as pd
import numpy as np

# Preprocessing
from sklearn.model_selection import train_test_split, StratifiedKFold, cross_val_score
from sklearn.preprocessing import StandardScaler, LabelEncoder
from sklearn.utils.class_weight import compute_class_weight

# Imbalanced data handling
from imblearn.over_sampling import SMOTE
from imblearn.under_sampling import RandomUnderSampler

# Machine learning models
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from lightgbm import LGBMClassifier
from xgboost import XGBClassifier
from catboost import CatBoostClassifier

# Evaluation
from sklearn.metrics import (
    precision_score,
    recall_score,
    balanced_accuracy_score,
    confusion_matrix,
    precision_recall_curve,
    auc,
    average_precision_score
)

# Suppress warnings
import warnings
warnings.filterwarnings('ignore')

In [346]:
df = pd.read_csv('data/medication_adherence.csv')

In [347]:
df.head()

,patient_id,age,gender,chronic_condition,medication_type,dosage,comorbidities,BMI,event_id,timestamp,...,blood_pressure,blood_glucose,activity_level,sleep_hours,stress_level,reminder_sent,reminder_response_time,social_support,location,future_non_adherence
0,P0103,45,M,diabetes,Metformin,50,0,24.800,E00000,2025-05-22 02:46:32,...,139/81,170.200,8030,5.200,3,1,15.000,0,work,0
1,P0436,79,M,hypertension,Metformin,500,0,25.200,E00001,2025-03-02 22:26:25,...,123/82,80.100,5338,8.800,2,1,94.000,0,home,0
2,P0861,71,F,diabetes,Salbutamol,20,3,28.600,E00002,2025-09-13 02:37:54,...,138/80,110.100,9021,5.600,4,1,86.000,2,hospital,0
3,P0271,89,M,hypertension,Atorvastatin,10,0,22.900,E00003,2025-01-26 03:54:55,...,122/84,157.200,4930,4.800,4,1,100.000,4,hospital,0
4,P0107,84,F,diabetes,Amlodipine,10,3,34.500,E00004,2025-02-20 03:28:37,...,125/79,147.300,5960,7.600,3,0,NaN,2,hospital,0


In [348]:
df.tail()

,patient_id,age,gender,chronic_condition,medication_type,dosage,comorbidities,BMI,event_id,timestamp,...,blood_pressure,blood_glucose,activity_level,sleep_hours,stress_level,reminder_sent,reminder_response_time,social_support,location,future_non_adherence
2795,P0151,79,F,cardiovascular,Atorvastatin,50,3,24.200,E02795,2025-01-07 05:26:36,...,125/80,164.100,3578,6.600,4,1,95.000,1,hospital,0
2796,P0384,59,M,diabetes,Atorvastatin,50,1,24.700,E02796,2025-07-31 15:52:22,...,122/77,149.700,2778,6.700,2,0,NaN,2,home,0
2797,P0638,66,F,asthma,Metformin,50,1,32.000,E02797,2025-02-12 08:02:04,...,119/89,162.400,1620,4.200,4,1,8.000,2,home,0
2798,P0813,54,M,hypertension,Metformin,100,2,19.400,E02798,2025-08-03 01:27:26,...,127/70,124.400,5054,8.800,3,1,33.000,5,work,0
2799,P0308,77,F,cardiovascular,Salbutamol,100,0,23.400,E02799,2025-03-22 20:24:04,...,118/83,73.700,8466,8.000,1,0,NaN,2,hospital,1


In [349]:
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2800 entries, 0 to 2799
Data columns (total 24 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   patient_id              2800 non-null   object 
 1   age                     2800 non-null   int64  
 2   gender                  2800 non-null   object 
 3   chronic_condition       2800 non-null   object 
 4   medication_type         2800 non-null   object 
 5   dosage                  2800 non-null   int64  
 6   comorbidities           2800 non-null   int64  
 7   BMI                     2800 non-null   float64
 8   event_id                2800 non-null   object 
 9   timestamp               2800 non-null   object 
 10  taken                   2800 non-null   int64  
 11  missed_reason           407 non-null    object 
 12  device_id               2800 non-null   object 
 13  heart_rate              2800 non-null   int64  
 14  blood_pressure          2800 non-null   

In [350]:
df.describe()

,age,dosage,comorbidities,BMI,taken,heart_rate,blood_glucose,activity_level,sleep_hours,stress_level,reminder_sent,reminder_response_time,social_support,future_non_adherence
count,2800.000,2800.000,2800.000,2800.000,2800.000,2800.000,2800.000,2800.000,2800.000,2800.000,2800.000,1966.000,2800.000,2800.000
mean,59.540,109.005,1.512,26.469,0.807,79.454,125.762,5434.715,6.492,2.990,0.702,60.833,2.476,0.249
std,17.216,169.817,1.126,4.885,0.394,11.456,31.714,2599.331,1.445,1.395,0.457,33.500,1.736,0.433
min,30.000,5.000,0.000,18.000,0.000,60.000,70.000,1001.000,4.000,1.000,0.000,5.000,0.000,0.000
25%,45.000,10.000,1.000,22.300,1.000,70.000,98.500,3191.500,5.300,2.000,0.000,31.000,1.000,0.000
50%,59.000,35.000,1.000,26.500,1.000,79.000,126.800,5405.500,6.500,3.000,1.000,60.000,2.000,0.000
75%,74.000,100.000,3.000,30.625,1.000,90.000,152.900,7664.500,7.700,4.000,1.000,90.000,4.000,0.000
max,89.000,500.000,3.000,35.000,1.000,99.000,180.000,9999.000,9.000,5.000,1.000,119.000,5.000,1.000


# Pre-processing

In [351]:
column_to_drop = [
    "patient_id",
    "event_id",
    "device_id",
    "timestamp"]

df = df.drop(columns=column_to_drop)
df.head()

,age,gender,chronic_condition,medication_type,dosage,comorbidities,BMI,taken,missed_reason,heart_rate,blood_pressure,blood_glucose,activity_level,sleep_hours,stress_level,reminder_sent,reminder_response_time,social_support,location,future_non_adherence
0,45,M,diabetes,Metformin,50,0,24.800,1,NaN,70,139/81,170.200,8030,5.200,3,1,15.000,0,work,0
1,79,M,hypertension,Metformin,500,0,25.200,0,forgetfulness,72,123/82,80.100,5338,8.800,2,1,94.000,0,home,0
2,71,F,diabetes,Salbutamol,20,3,28.600,1,NaN,97,138/80,110.100,9021,5.600,4,1,86.000,2,hospital,0
3,89,M,hypertension,Atorvastatin,10,0,22.900,1,NaN,68,122/84,157.200,4930,4.800,4,1,100.000,4,hospital,0
4,84,F,diabetes,Amlodipine,10,3,34.500,1,NaN,75,125/79,147.300,5960,7.600,3,0,NaN,2,hospital,0


In [352]:
df = df.drop(['missed_reason', 'taken'], axis=1)

In [353]:
# Split blood_pressure
df[['systolic', 'diastolic']] = df['blood_pressure'].str.split('/', expand=True)

# Ubah ke integer
df['systolic'] = df['systolic'].astype(int)
df['diastolic'] = df['diastolic'].astype(int)

# Drop kolom lama
df = df.drop('blood_pressure', axis=1)

In [354]:
df.head()


,age,gender,chronic_condition,medication_type,dosage,comorbidities,BMI,heart_rate,blood_glucose,activity_level,sleep_hours,stress_level,reminder_sent,reminder_response_time,social_support,location,future_non_adherence,systolic,diastolic
0,45,M,diabetes,Metformin,50,0,24.800,70,170.200,8030,5.200,3,1,15.000,0,work,0,139,81
1,79,M,hypertension,Metformin,500,0,25.200,72,80.100,5338,8.800,2,1,94.000,0,home,0,123,82
2,71,F,diabetes,Salbutamol,20,3,28.600,97,110.100,9021,5.600,4,1,86.000,2,hospital,0,138,80
3,89,M,hypertension,Atorvastatin,10,0,22.900,68,157.200,4930,4.800,4,1,100.000,4,hospital,0,122,84
4,84,F,diabetes,Amlodipine,10,3,34.500,75,147.300,5960,7.600,3,0,NaN,2,hospital,0,125,79


In [355]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2800 entries, 0 to 2799
Data columns (total 19 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   age                     2800 non-null   int64  
 1   gender                  2800 non-null   object 
 2   chronic_condition       2800 non-null   object 
 3   medication_type         2800 non-null   object 
 4   dosage                  2800 non-null   int64  
 5   comorbidities           2800 non-null   int64  
 6   BMI                     2800 non-null   float64
 7   heart_rate              2800 non-null   int64  
 8   blood_glucose           2800 non-null   float64
 9   activity_level          2800 non-null   int64  
 10  sleep_hours             2800 non-null   float64
 11  stress_level            2800 non-null   int64  
 12  reminder_sent           2800 non-null   int64  
 13  reminder_response_time  1966 non-null   float64
 14  social_support          2800 non-null   

In [356]:
df.isnull().sum()

age                         0
gender                      0
chronic_condition           0
medication_type             0
dosage                      0
comorbidities               0
BMI                         0
heart_rate                  0
blood_glucose               0
activity_level              0
sleep_hours                 0
stress_level                0
reminder_sent               0
reminder_response_time    834
social_support              0
location                    0
future_non_adherence        0
systolic                    0
diastolic                   0
dtype: int64

In [357]:
pd.crosstab(
    df['reminder_sent'],
    df['reminder_response_time'].isnull()
)

reminder_response_time,False,True
reminder_sent,,
0,0,834
1,1966,0


In [358]:
df['reminder_response_time'] = df['reminder_response_time'].fillna(0)

In [359]:
df.isnull().sum()

age                       0
gender                    0
chronic_condition         0
medication_type           0
dosage                    0
comorbidities             0
BMI                       0
heart_rate                0
blood_glucose             0
activity_level            0
sleep_hours               0
stress_level              0
reminder_sent             0
reminder_response_time    0
social_support            0
location                  0
future_non_adherence      0
systolic                  0
diastolic                 0
dtype: int64

In [360]:
X = df.drop('future_non_adherence', axis=1)
y = df['future_non_adherence']

In [361]:
y.value_counts()

future_non_adherence
0    2102
1     698
Name: count, dtype: int64

In [362]:
X.select_dtypes(include='object').columns

Index(['gender', 'chronic_condition', 'medication_type', 'location'], dtype='object')

In [363]:
X['medication_type'].nunique()

4

In [364]:
X_encoded = pd.get_dummies(
    X,
    columns=['gender', 'chronic_condition', 'medication_type', 'location'],
    drop_first=True
)

X_encoded.head()

,age,dosage,comorbidities,BMI,heart_rate,blood_glucose,activity_level,sleep_hours,stress_level,reminder_sent,...,diastolic,gender_M,chronic_condition_cardiovascular,chronic_condition_diabetes,chronic_condition_hypertension,medication_type_Atorvastatin,medication_type_Metformin,medication_type_Salbutamol,location_hospital,location_work
0,45,50,0,24.800,70,170.200,8030,5.200,3,1,...,81,True,False,True,False,False,True,False,False,True
1,79,500,0,25.200,72,80.100,5338,8.800,2,1,...,82,True,False,False,True,False,True,False,False,False
2,71,20,3,28.600,97,110.100,9021,5.600,4,1,...,80,False,False,True,False,False,False,True,True,False
3,89,10,0,22.900,68,157.200,4930,4.800,4,1,...,84,True,False,False,True,True,False,False,True,False
4,84,10,3,34.500,75,147.300,5960,7.600,3,0,...,79,False,False,True,False,False,False,False,True,False


In [365]:
X_encoded.shape

(2800, 23)

In [366]:
bool_cols = X_encoded.select_dtypes(include='bool').columns
X_encoded[bool_cols] = X_encoded[bool_cols].astype(int)

# Gunakan X_encoded sebagai input
X_for_cv = X_encoded.copy()

In [367]:
X_encoded.head()

,age,dosage,comorbidities,BMI,heart_rate,blood_glucose,activity_level,sleep_hours,stress_level,reminder_sent,...,diastolic,gender_M,chronic_condition_cardiovascular,chronic_condition_diabetes,chronic_condition_hypertension,medication_type_Atorvastatin,medication_type_Metformin,medication_type_Salbutamol,location_hospital,location_work
0,45,50,0,24.800,70,170.200,8030,5.200,3,1,...,81,1,0,1,0,0,1,0,0,1
1,79,500,0,25.200,72,80.100,5338,8.800,2,1,...,82,1,0,0,1,0,1,0,0,0
2,71,20,3,28.600,97,110.100,9021,5.600,4,1,...,80,0,0,1,0,0,0,1,1,0
3,89,10,0,22.900,68,157.200,4930,4.800,4,1,...,84,1,0,0,1,1,0,0,1,0
4,84,10,3,34.500,75,147.300,5960,7.600,3,0,...,79,0,0,1,0,0,0,0,1,0


In [368]:
X_encoded.shape

(2800, 23)

# BASELINE 

In [369]:
models = {

    "Logistic Regression": LogisticRegression(
        max_iter=1000,
        random_state=42
    ),

    "SVM (RBF)": SVC(
        kernel='rbf',
        probability=True,
        random_state=42
    ),

    "LightGBM": LGBMClassifier(
        random_state=42
    ),

    "XGBoost": XGBClassifier(
        use_label_encoder=False,
        eval_metric='logloss',
        random_state=42
    ),

    "CatBoost": CatBoostClassifier(
        verbose=0,
        random_state=42
    )
}

results = []

In [370]:
skf = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)

results = []

for name, model in models.items():

    recall_list = []
    precision_list = []
    specificity_list = []
    balanced_acc_list = []
    pr_auc_list = []

    for train_index, test_index in skf.split(X_encoded, y):

        X_tr = X_encoded.iloc[train_index]
        X_te = X_encoded.iloc[test_index]
        y_tr = y.iloc[train_index]
        y_te = y.iloc[test_index]

        model.fit(X_tr, y_tr)

        y_pred = model.predict(X_te)
        y_proba = model.predict_proba(X_te)[:, 1]

        tn, fp, fn, tp = confusion_matrix(y_te, y_pred).ravel()

        recall_list.append(recall_score(y_te, y_pred))
        precision_list.append(precision_score(y_te, y_pred, zero_division=0))
        specificity_list.append(tn / (tn + fp))
        balanced_acc_list.append(balanced_accuracy_score(y_te, y_pred))

        precision_curve, recall_curve, _ = precision_recall_curve(y_te, y_proba)
        pr_auc_list.append(auc(recall_curve, precision_curve))

    results.append([
        name,
        np.mean(recall_list),
        np.mean(precision_list),
        np.mean(specificity_list),
        np.mean(balanced_acc_list),
        np.mean(pr_auc_list)
    ])

[LightGBM] [Info] Number of positive: 559, number of negative: 1681
[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 0.000233 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 1046
[LightGBM] [Info] Number of data points in the train set: 2240, number of used features: 23
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.249554 -> initscore=-1.100995
[LightGBM] [Info] Start training from score -1.100995
[LightGBM] [Info] Number of positive: 559, number of negative: 1681
[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 0.000179 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 1046
[LightGBM] [Info] Number of data points in the train set: 2240, number of used features: 23
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.249554 -> initscore=-1.100995
[LightGBM] [Info] Start training from score -1.100995
[LightGBM] [Info] Nu

In [371]:
baseline_df = pd.DataFrame(results, columns=[
    "Model",
    "Sensitivity (Recall)",
    "Precision",
    "Specificity",
    "Balanced Accuracy",
    "PR-AUC"
])

baseline_df.sort_values(by="PR-AUC", ascending=False)

,Model,Sensitivity (Recall),Precision,Specificity,Balanced Accuracy,PR-AUC
4,CatBoost,0.009,0.320,0.995,0.502,0.273
2,LightGBM,0.057,0.277,0.951,0.504,0.262
0,Logistic Regression,0.000,0.000,1.000,0.500,0.261
3,XGBoost,0.079,0.262,0.923,0.501,0.252
1,SVM (RBF),0.000,0.000,1.000,0.500,0.247


Hasil baseline menunjukkan bahwa seluruh model cenderung bias terhadap kelas mayoritas, dengan nilai sensitivity yang sangat rendah (0.000–0.079) dan specificity yang sangat tinggi (0.923–1.000). Hal ini mengindikasikan bahwa model lebih memilih meminimalkan kesalahan pada kelas mayoritas dengan mengorbankan kemampuan mendeteksi pasien non-adherence. Balanced accuracy yang mendekati 0.5 menguatkan bahwa performa model pada kondisi tidak seimbang masih belum optimal. Meskipun demikian, nilai PR-AUC yang relatif stabil menunjukkan bahwa model masih memiliki kemampuan diskriminatif pada tingkat probabilitas, sehingga intervensi teknik imbalance learning diperlukan untuk meningkatkan sensitivity tanpa mengorbankan stabilitas model secara signifikan.

# Cost-sensitive learning 

In [372]:
classes = np.unique(y)
class_weights = compute_class_weight(
    class_weight='balanced',
    classes=classes,
    y=y
)

class_weight_dict = dict(zip(classes, class_weights))

print(class_weight_dict)

{np.int64(0): np.float64(0.6660323501427212), np.int64(1): np.float64(2.005730659025788)}


In [373]:
models_cs = {

    "Logistic Regression": LogisticRegression(
        class_weight=class_weight_dict,
        max_iter=1000,
        random_state=42
    ),

    "SVM (RBF)": SVC(
        kernel='rbf',
        class_weight=class_weight_dict,
        probability=True,
        random_state=42
    ),

    "LightGBM": LGBMClassifier(
        class_weight=class_weight_dict,
        random_state=42
    ),

    "XGBoost": XGBClassifier(
        scale_pos_weight=class_weight_dict[1] / class_weight_dict[0],
        use_label_encoder=False,
        eval_metric='logloss',
        random_state=42
    ),

    "CatBoost": CatBoostClassifier(
        class_weights=[class_weight_dict[0], class_weight_dict[1]],
        verbose=0,
        random_state=42
    )
}

In [374]:
skf = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)

results_cs = []

for name, model in models_cs.items():

    recall_list = []
    precision_list = []
    specificity_list = []
    balanced_acc_list = []
    pr_auc_list = []

    for train_index, test_index in skf.split(X_encoded, y):

        X_tr = X_encoded.iloc[train_index]
        X_te = X_encoded.iloc[test_index]
        y_tr = y.iloc[train_index]
        y_te = y.iloc[test_index]

        model.fit(X_tr, y_tr)

        y_pred = model.predict(X_te)
        y_proba = model.predict_proba(X_te)[:, 1]

        tn, fp, fn, tp = confusion_matrix(y_te, y_pred).ravel()

        recall_list.append(recall_score(y_te, y_pred))
        precision_list.append(precision_score(y_te, y_pred, zero_division=0))
        specificity_list.append(tn / (tn + fp))
        balanced_acc_list.append(balanced_accuracy_score(y_te, y_pred))

        precision_curve, recall_curve, _ = precision_recall_curve(y_te, y_proba)
        pr_auc_list.append(auc(recall_curve, precision_curve))

    results_cs.append([
        name,
        np.mean(recall_list),
        np.mean(precision_list),
        np.mean(specificity_list),
        np.mean(balanced_acc_list),
        np.mean(pr_auc_list)
    ])

[LightGBM] [Info] Number of positive: 559, number of negative: 1681
[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 0.000213 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 1046
[LightGBM] [Info] Number of data points in the train set: 2240, number of used features: 23
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.500358 -> initscore=0.001431
[LightGBM] [Info] Start training from score 0.001431
[LightGBM] [Info] Number of positive: 559, number of negative: 1681
[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 0.000171 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 1046
[LightGBM] [Info] Number of data points in the train set: 2240, number of used features: 23
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.500358 -> initscore=0.001431
[LightGBM] [Info] Start training from score 0.001431
[LightGBM] [Info] Number

In [375]:
cost_sensitive_df = pd.DataFrame(results_cs, columns=[
    "Model",
    "Sensitivity (Recall)",
    "Precision",
    "Specificity",
    "Balanced Accuracy",
    "PR-AUC"
])

cost_sensitive_df.sort_values(by="PR-AUC", ascending=False)

,Model,Sensitivity (Recall),Precision,Specificity,Balanced Accuracy,PR-AUC
4,CatBoost,0.171,0.282,0.855,0.513,0.268
2,LightGBM,0.208,0.267,0.814,0.511,0.265
3,XGBoost,0.169,0.286,0.861,0.515,0.262
0,Logistic Regression,0.489,0.254,0.522,0.505,0.255
1,SVM (RBF),0.324,0.231,0.633,0.478,0.249


Hasil eksperimen menunjukkan bahwa pendekatan cost-sensitive learning secara signifikan meningkatkan sensitivity pada seluruh model dibandingkan baseline. Logistic Regression menunjukkan peningkatan paling drastis dari 0.000 menjadi 0.489, menandakan bahwa penalti kesalahan pada kelas minoritas efektif mendorong model linear untuk mendeteksi pasien non-adherence. Pada model berbasis boosting seperti LightGBM, XGBoost, dan CatBoost, peningkatan recall bersifat moderat, mengindikasikan bahwa model ensemble modern memiliki ketahanan awal terhadap ketidakseimbangan kelas. Secara keseluruhan, pendekatan cost-sensitive berhasil mengurangi bias terhadap kelas mayoritas tanpa menurunkan PR-AUC secara signifikan.

# SMOTE

In [376]:
smote = SMOTE(random_state=42)

X_train_smote, y_train_smote = smote.fit_resample(X_train, y_train)

print("Distribusi sebelum SMOTE:", Counter(y_train))
print("Distribusi setelah SMOTE:", Counter(y_train_smote))

Distribusi sebelum SMOTE: Counter({0: 1682, 1: 558})
Distribusi setelah SMOTE: Counter({0: 1682, 1: 1682})


In [377]:
models_smote = {
    "Logistic Regression": LogisticRegression(max_iter=1000, random_state=42),
    "SVM (RBF)": SVC(kernel='rbf', probability=True, random_state=42),
    "LightGBM": LGBMClassifier(random_state=42),
    "XGBoost": XGBClassifier(use_label_encoder=False, eval_metric='logloss', random_state=42),
    "CatBoost": CatBoostClassifier(verbose=0, random_state=42)
}

In [378]:
results_smote = []

for name, model in models_smote.items():
    model.fit(X_train_smote, y_train_smote)
    y_pred = model.predict(X_test)
    y_proba = model.predict_proba(X_test)[:, 1]

    recall = recall_score(y_test, y_pred)
    precision = precision_score(y_test, y_pred)
    specificity = recall_score(y_test, y_pred, pos_label=0)
    bal_acc = balanced_accuracy_score(y_test, y_pred)
    pr_auc = average_precision_score(y_test, y_proba)

    results_smote.append([
        name, recall, precision, specificity, bal_acc, pr_auc
    ])

df_smote = pd.DataFrame(results_smote, columns=[
    "Model", "Sensitivity (Recall)", "Precision",
    "Specificity", "Balanced Accuracy", "PR-AUC"
])

df_smote.sort_values(by="PR-AUC", ascending=False)

[LightGBM] [Info] Number of positive: 1682, number of negative: 1682
[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 0.000249 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 1532
[LightGBM] [Info] Number of data points in the train set: 3364, number of used features: 23
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.500000 -> initscore=0.000000


,Model,Sensitivity (Recall),Precision,Specificity,Balanced Accuracy,PR-AUC
0,Logistic Regression,0.271,0.255,0.736,0.504,0.271
3,XGBoost,0.164,0.267,0.850,0.507,0.261
4,CatBoost,0.129,0.243,0.867,0.498,0.250
2,LightGBM,0.100,0.194,0.862,0.481,0.239
1,SVM (RBF),0.393,0.225,0.550,0.471,0.229


Cost-sensitive learning lebih efektif dibanding SMOTE pada tree-based boosting methods, sementara SMOTE hanya memberikan perbaikan terbatas pada linear classifiers seperti Logistic Regression.

# Random Under Sampling

In [ ]:
models = {
    "Logistic Regression": LogisticRegression(max_iter=1000, random_state=42),
    "SVM (RBF)": SVC(kernel='rbf', probability=True, random_state=42),
    "LightGBM": LGBMClassifier(random_state=42),
    "XGBoost": XGBClassifier(use_label_encoder=False, eval_metric='logloss', random_state=42),
    "CatBoost": CatBoostClassifier(verbose=0, random_state=42)
}

In [381]:
rus = RandomUnderSampler(random_state=42)

X_train_rus, y_train_rus = rus.fit_resample(X_train, y_train)

print("Distribusi setelah RUS:")
print(y_train_rus.value_counts())

Distribusi setelah RUS:
future_non_adherence
0    558
1    558
Name: count, dtype: int64


In [382]:
results_rus = []

for name, model in models.items():
    
    model.fit(X_train_rus, y_train_rus)
    
    y_pred = model.predict(X_test)
    y_proba = model.predict_proba(X_test)[:, 1]
    
    cm = confusion_matrix(y_test, y_pred)
    tn, fp, fn, tp = cm.ravel()
    
    recall = recall_score(y_test, y_pred)
    precision = precision_score(y_test, y_pred, zero_division=0)
    specificity = tn / (tn + fp) if (tn + fp) > 0 else 0
    balanced_acc = balanced_accuracy_score(y_test, y_pred)
    pr_auc = average_precision_score(y_test, y_proba)
    
    results_rus.append([
        name,
        recall,
        precision,
        specificity,
        balanced_acc,
        pr_auc
    ])

df_rus = pd.DataFrame(results_rus, columns=[
    "Model", 
    "Sensitivity (Recall)", 
    "Precision", 
    "Specificity", 
    "Balanced Accuracy", 
    "PR-AUC"
])

df_rus.sort_values(by="PR-AUC", ascending=False)

[LightGBM] [Info] Number of positive: 558, number of negative: 558
[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 0.000198 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 1033
[LightGBM] [Info] Number of data points in the train set: 1116, number of used features: 23
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.500000 -> initscore=0.000000


,Model,Sensitivity (Recall),Precision,Specificity,Balanced Accuracy,PR-AUC
1,SVM (RBF),0.279,0.227,0.683,0.481,0.250
0,Logistic Regression,0.457,0.238,0.512,0.485,0.242
3,XGBoost,0.443,0.215,0.462,0.452,0.242
4,CatBoost,0.393,0.200,0.476,0.435,0.221
2,LightGBM,0.471,0.229,0.471,0.471,0.218


Cost-sensitive boosting lebih superior dibanding data-level under-sampling.

# Combine Result Cost Sensitive Learning, SMOTE & RUS

In [383]:
columns = [
    "Model",
    "Sensitivity (Recall)",
    "Precision",
    "Specificity",
    "Balanced Accuracy",
    "PR-AUC"
]

In [388]:
baseline_df["Teknik"] = "Baseline"
cost_sensitive_df["Teknik"] = "Cost-Sensitive"
df_smote["Teknik"] = "SMOTE"
df_rus["Teknik"] = "RUS"

In [390]:
comparison_df = pd.concat(
    [baseline_df, cost_sensitive_df, df_smote, df_rus],
    ignore_index=True
)

In [394]:
comparison_df = comparison_df.sort_values(
    by="PR-AUC",
    ascending=False
).reset_index(drop=True)

comparison_df

,Model,Sensitivity (Recall),Precision,Specificity,Balanced Accuracy,PR-AUC,Teknik
0,CatBoost,0.009,0.320,0.995,0.502,0.273,Baseline
1,Logistic Regression,0.271,0.255,0.736,0.504,0.271,SMOTE
2,CatBoost,0.171,0.282,0.855,0.513,0.268,Cost-Sensitive
3,LightGBM,0.208,0.267,0.814,0.511,0.265,Cost-Sensitive
4,LightGBM,0.057,0.277,0.951,0.504,0.262,Baseline
5,XGBoost,0.169,0.286,0.861,0.515,0.262,Cost-Sensitive
6,Logistic Regression,0.000,0.000,1.000,0.500,0.261,Baseline
7,XGBoost,0.164,0.267,0.850,0.507,0.261,SMOTE
8,Logistic Regression,0.489,0.254,0.522,0.505,0.255,Cost-Sensitive
9,XGBoost,0.079,0.262,0.923,0.501,0.252,Baseline


In [396]:
pivot_pr_auc = comparison_df.pivot_table(
    index="Model",
    columns="Teknik",
    values="PR-AUC"
)

pivot_pr_auc

Teknik,Baseline,Cost-Sensitive,RUS,SMOTE
Model,,,,
CatBoost,0.273,0.268,0.221,0.250
LightGBM,0.262,0.265,0.218,0.239
Logistic Regression,0.261,0.255,0.242,0.271
SVM (RBF),0.247,0.249,0.250,0.229
XGBoost,0.252,0.262,0.242,0.261
